universe/Universe.h

```diff
@@ -399,13 +399,17 @@ class FO_COMMON_API Universe {
 
     /** Set items unlocked before turn 1 from \p future.*/
     void SetInitiallyUnlockedItems(Pending::Pending<std::vector<ItemSpec>>&& future);
+    void SetInitiallyUnlockedItemsPrewarp(Pending::Pending<std::vector<ItemSpec>>&& future);
```

Comment:
Are extra and specific C++ functions for this necessary? This is overly specific to a particular scenario you are creating and would preferably be handled just with Python.

Reply:
Well, the very first thing the ServerApp constructor (see server/ServerApp.cpp) does, long before anything like GalaxySetupData might become available, is calling ServerApp::StartBackgroundParsing(), and that already contains the hardcoded file names "default/scripting/starting_unlocks/*.inf" right there is C++ code.  But the pre-warp era clearly needs different files, so with the current code architecture, i see no choice but to load both sets of files, with files names and eras hardcoded in a subroutine of the ServerApp constructor, and then of course era names also hardcoded in Universe member names because some place is needed to store the data that was just parsed.

Of course, one could totally change the architecture: stop using Pending::Pending for these three files and let the parsing be initiated much later, synchronously, when the universe generator actually uses the data from these files: this data is used at exactly one place, once, in file default/python/universe_generation/empires.py, function setup_empire(), and then never looked at again.  Are you telling me to make that architectural change?

If so, it might make sense to do the parsing directly in Python in an ad hoc manner rather than calling the heavy-weight and cumbersome C++/Boost parsing routines.  After all, nothing from the files is actually used except the raw names.  That would probably be much simpler and likely also faster (though see above, speed almost certainly doesn't matter here).  

I do think it is silly to have these three files parsed asynchronously.  FreeOrion startup is certainly slow - about half a minute on my notebook, while physically copying the complete default/scripting/ directory to another partition takes less than a second, so reading the data from disk is not what makes it slow - so using concurrency for these tiny files is definitely premature optimization and more likely adding overhead and needless complexity than making anything faster or simpler.

Or how do you want this done?

Reply:
The parsing in general can be very slow, so doing it in another thread isn't a premature optimization; this functionality was implemented to resolve a specific and substantial delay when doing it all single-threaded. Doing it for those particular files is probably unnecessary, but judging whether to do it in different ways for different files would be unnecessarily complicated; it's easier, and not much cost, to just do them all asynchronously.

That said, I don't follow your argument... The slowness certainly isn't the disk read, it's the actual parsing / processing the data. But that's not the point; the issue is the actual parsing of (mostly bigger and more complicated) scripts, which isn't needed immediately on startup, and can be done in parallel with the rest of the initialization to avoid even longer waits for the UI to become usable.

Regardless, for this case, I think just reading the scripts in Python would make more sense. If you want to implement that, please do, though whether it should happen really depends on your level of interest and motivation. I'd also prefer to avoid the extra overly-specific new lists of content in the C++ code, but I won't make that a requirement for acceptance of these changes, as it can be fixed later when someone is so inclined.

Reply:
I see your points, so i have implemented reading these three particular scripts in Python, see commit 0b7a2ea2.

What do you mean by "new lists of content in the C++ code"? I just looked through the correct version of my combined diff and couldn't find any.

Reply:
I meant the new lists for prewarp buildings and items.

Reply:
Oh, those...  I already deleted them in the same commit 0b7a2ea2 because they were no longer needed after moving the parsing to Python.  So i conclude that this conversation can also marked as "resolved".